src/nifreeze/model/pet.py

```diff
@@ -95,62 +125,114 @@ def __init__(
+                nb.Nifti1Image(data, self._dataset.affine), self._smooth_fwhm
+            )
+            data = smoothed_img.get_fdata()
+
+        if self._thresh_pct > 0:
+            thresh_val = np.percentile(data, self._thresh_pct)
+            data[data < thresh_val] = 0
 
-        self._datashape = None
-        self._mask = None
+        # Convert data into V (voxels) x T (timepoints)
+        return data.reshape((-1, data.shape[-1])) if brainmask is None else data[brainmask]
 
     @property
-    def is_fitted(self):
+    def is_fitted(self) -> bool:
         return self._locked_fit is not None
 
+    @abstractmethod
+    def fit_predict(self, index: int | None = None, **kwargs) -> Union[np.ndarray, None]:
+        """Predict the corrected volume."""
+        return None
+
+
+class BSplinePETModel(BasePETModel):
+    """A PET imaging realignment model based on B-Spline approximation."""
+
+    __slots__ = (
+        "_t",
+        "_order",
+        "_n_ctrl",
+    )
+
+    def __init__(
+        self,
+        dataset: PET,
+        n_ctrl: int | None = None,
+        order: int = 3,
+        **kwargs,
+    ):
+        """Create the B-Spline interpolating matrix.
+
+        Parameters
+        ----------
+        n_ctrl : :obj:`int`
+            Number of B-Spline control points. If `None`, then one control point every
+            six timepoints will be used. The less control points, the smoother is the
+            model.
+        order : :obj:`int`
+            Order of the B-Spline approximation.
+        """
+
+        super().__init__(dataset, **kwargs)
+
+        self._order = order
+
+        # Calculate index coordinates in the B-Spline grid
+        self._n_ctrl = n_ctrl or (len(self._x) // 4) + 1
+
+        # B-Spline knots
+        self._t = np.arange(-3, self._n_ctrl + 4, dtype="float32")
+
     def _fit(self, index: int | None = None, n_jobs=None, **kwargs) -> int:
         """Fit the model."""
 
+        n_jobs = n_jobs or min(cpu_count() or 1, 8)
+
         if self._locked_fit is not None:
             return n_jobs
 
         if index is not None:
             raise NotImplementedError("Fitting with held-out data is not supported")
-        timepoints = kwargs.get("timepoints", None) or self._x
-        x = (np.array(timepoints, dtype="float32") / self._xlim) * self._n_ctrl
 
-        data = self._dataset.dataobj
-        brainmask = self._dataset.brainmask
+        data = self._preproces_data()
 
-        if self._smooth_fwhm > 0:
-            smoothed_img = smooth_image(
-                nb.Nifti1Image(data, self._dataset.affine), self._smooth_fwhm
-            )
-            data = smoothed_img.get_fdata()
-
-        if self._thresh_pct > 0:
-            thresh_val = np.percentile(data, self._thresh_pct)
-            data[data < thresh_val] = 0
-
-        # Convert data into V (voxels) x T (timepoints)
-        data = data.reshape((-1, data.shape[-1])) if brainmask is None else data[brainmask]
+        # ToDo
```

Comment:
Allowing _fit to override timepoints through kwargs duplicates information that was already validated and stored on self._x during initialization. Dropping that extra kwarg simplifies the API and avoids inconsistent inputs

Reply:
Not sure I understand: maybe the reply is answering a question elsewhere?